src/compiler/utilities.ts

```diff
@@ -8587,140 +8587,233 @@ export function getSetExternalModuleIndicator(options: CompilerOptions): (file:
     }
 }
 
+type CompilerOptionKeys = keyof { [K in keyof CompilerOptions as string extends K ? never : K]: any; };
+function createComputedCompilerOptions<T extends Record<string, CompilerOptionKeys[]>>(
+    options: {
+        [K in keyof T & CompilerOptionKeys]: {
+            dependencies: T[K];
+            computeValue: (compilerOptions: Pick<CompilerOptions, K | T[K][number]>) => Exclude<CompilerOptions[K], undefined>;
+        };
+    },
+) {
+    return options;
+}
+
 /** @internal */
-export interface ComputedCompilerOption<N extends keyof CompilerOptions, K extends keyof CompilerOptions, R extends CompilerOptions[N]> {
-    name: N;
-    dependencies: readonly K[];
-    computeValue: (compilerOptions: Pick<CompilerOptions, N | K>) => R;
-}
-
-/** @internal */
-function computedCompilerOption<N extends keyof CompilerOptions, K extends keyof CompilerOptions, R extends CompilerOptions[N]>(
-    name: N,
-    dependencies: readonly K[],
-    computeValue: (compilerOptions: Pick<CompilerOptions, N | K>) => R,
-): ComputedCompilerOption<N, K, R> {
-    return { name, dependencies, computeValue };
-}
-
-/** @internal */
-export const computedOptions = {
-    target: computedCompilerOption("target", ["module"], compilerOptions => {
-        return compilerOptions.target ??
-            ((compilerOptions.module === ModuleKind.Node16 && ScriptTarget.ES2022) ||
-                (compilerOptions.module === ModuleKind.NodeNext && ScriptTarget.ESNext) ||
-                ScriptTarget.ES5);
-    }),
-    module: computedCompilerOption("module", ["target"], (compilerOptions): ModuleKind => {
-        return typeof compilerOptions.module === "number" ?
-            compilerOptions.module :
-            computedOptions.target.computeValue(compilerOptions) >= ScriptTarget.ES2015 ? ModuleKind.ES2015 : ModuleKind.CommonJS;
-    }),
-    moduleResolution: computedCompilerOption("moduleResolution", ["module", "target"], (compilerOptions): ModuleResolutionKind => {
-        let moduleResolution = compilerOptions.moduleResolution;
-        if (moduleResolution === undefined) {
+export const computedOptions = createComputedCompilerOptions({

```

Comment:
Thanks @Andarist for helping with this tricky inference 🌟 

This only works in recent nightlies, hence the “LKG” bump (cc @jakebailey)

Reply:
I don't mind bumping the TS version, though I am wondering which PR was needed to make this pattern work (and of course if there's any change at all that we could avoid it 😅).

But yeah, this pattern is nicer than the previous IMO.